systemdspawner/systemdspawner.py

```diff
@@ -147,25 +149,37 @@ class SystemdSpawner(Spawner):
         """
     ).tag(config=True)
 
+    # Characters that are safe for systemd units.
+    safe_chars = set(string.ascii_lowercase + string.digits + string.ascii_uppercase + '_-')
+
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         # All traitlets configurables are configured by now
         self.unit_name = self._expand_user_vars(self.unit_name_template)
 
         self.log.debug('user:%s Initialized spawner with unit %s', self.user.name, self.unit_name)
 
-    def _expand_user_vars(self, string):
+    def _escape_variable(self, in_string):
+        """
+        Escape variables for systemd unit naming
+        """
+        return escapism.escape(in_string, safe=self.safe_chars, escape_char='-')
```

Comment:
A different `escape_char` than a dash, maybe a colon, since it is legal in unit names but illegal in usernames, would be easier if one were to disassemble the unit name for other purposes (e.g. monitoring of running units).

Legal characters in units names are (from `man systemd.unit`):
> (ASCII letters, digits, ":", "-", "_", ".", and "\")

Also, the resulting unit name may not be longer than 248 letters, since `.service` will be added and the total may not exceed 256 letters. I think this is relevant, since people can freely choose the name of the server.